plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarConnectorManagedLedgerFactory.java

```diff
@@ -0,0 +1,200 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.pulsar;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Strings;
+import com.google.common.collect.ImmutableMap;
+import io.airlift.log.Logger;
+import org.apache.bookkeeper.common.util.OrderedScheduler;
+import org.apache.bookkeeper.conf.ClientConfiguration;
+import org.apache.bookkeeper.mledger.LedgerOffloader;
+import org.apache.bookkeeper.mledger.LedgerOffloaderFactory;
+import org.apache.bookkeeper.mledger.ManagedLedgerConfig;
+import org.apache.bookkeeper.mledger.ManagedLedgerFactory;
+import org.apache.bookkeeper.mledger.ManagedLedgerFactoryConfig;
+import org.apache.bookkeeper.mledger.impl.ManagedLedgerFactoryImpl;
+import org.apache.bookkeeper.mledger.impl.NullLedgerOffloader;
+import org.apache.bookkeeper.mledger.offload.Offloaders;
+import org.apache.bookkeeper.mledger.offload.OffloadersCache;
+import org.apache.bookkeeper.stats.StatsProvider;
+import org.apache.pulsar.PulsarVersion;
+import org.apache.pulsar.metadata.api.MetadataStore;
+import org.apache.pulsar.metadata.api.MetadataStoreConfig;
+import org.apache.pulsar.metadata.api.MetadataStoreFactory;
+import org.apache.pulsar.shade.org.apache.pulsar.common.naming.NamespaceName;
+import org.apache.pulsar.shade.org.apache.pulsar.common.policies.data.OffloadPolicies;
+import org.apache.pulsar.shade.org.apache.pulsar.common.policies.data.OffloadPoliciesImpl;
+
+import javax.inject.Inject;
+
+import java.io.IOException;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+public class PulsarConnectorManagedLedgerFactory
+        implements PulsarConnectorCache
+{
+    private static final Logger log = Logger.get(PulsarConnectorManagedLedgerFactory.class);
+
+    @VisibleForTesting
+    static PulsarConnectorCache instance;
+
+    private final MetadataStore metadataStore;
+    private final ManagedLedgerFactory managedLedgerFactory;
+
+    private final StatsProvider statsProvider;
+    private OrderedScheduler offloaderScheduler;
+    private OffloadersCache offloadersCache = new OffloadersCache();
+    private LedgerOffloader defaultOffloader;
+    private Map<NamespaceName, LedgerOffloader> offloaderMap = new ConcurrentHashMap<>();
+
+    @Inject
+    public PulsarConnectorManagedLedgerFactory(PulsarConnectorConfig pulsarConnectorConfig) throws Exception
+    {
+        metadataStore = MetadataStoreFactory.create(pulsarConnectorConfig.getZookeeperUri(),
+                MetadataStoreConfig.builder().build());
+        managedLedgerFactory = initManagedLedgerFactory(pulsarConnectorConfig);
+        statsProvider = PulsarConnectorUtils.createInstance(pulsarConnectorConfig.getStatsProvider(),
```

Comment:
previously we want to able to load different statsprovider, but since we switched to JMX based implementation to align with project convention, we don't need it anymore, removed.